File: api/src/timeline/merger.ts
import { Kafka, Consumer, EachMessagePayload, Producer } from 'kafkajs';

interface TimelineEvent {
  id: string;
  sessionId: string;
  pageId?: string;
  ts: number;
  tsPage?: number;
  type: string;
  subtype?: string;
  payload: any;
  seq: number;
}

const SOURCE_TOPICS = [
  'obs.network.requests',
  'obs.network.responses',
  'obs.dom.mutations',
  'obs.js.events',
  'obs.storage.events',
  'obs.screenshots',
  'obs.performance',
];

const TYPE_MAP: Record<string, string> = {
  'obs.network.requests': 'network',
  'obs.network.responses': 'network',
  'obs.dom.mutations': 'dom',
  'obs.js.events': 'js',
  'obs.storage.events': 'storage',
  'obs.screenshots': 'screenshot',
  'obs.performance': 'performance',
};

export class TimelineMerger {
  private consumer: Consumer;
  private producer: Producer;
  private seqMap = new Map<string, { seq: number; lastAccess: number }>();
  private cleanupTimer: ReturnType<typeof setInterval>;

  constructor(kafka: Kafka, groupId: string) {
    this.consumer = kafka.consumer({ groupId });
    this.producer = kafka.producer();
    this.cleanupTimer = setInterval(() => this.evictStale(), 5 * 60 * 1000);
  }

  async start(): Promise<void> {
    await this.consumer.connect();
    await this.producer.connect();

    for (const topic of SOURCE_TOPICS) {
      await this.consumer.subscribe({ topic });
    }

    await this.consumer.run({
      eachMessage: async (payload: EachMessagePayload) => this.process(payload),
    });

    console.log('[TimelineMerger] Started');
  }

  private async process({ topic, message }: EachMessagePayload): Promise<void> {
    try {
      const raw = JSON.parse(message.value!.toString());
      const sessionId = raw.sessionId ?? message.key?.toString();
      if (!sessionId) return;

      const entry = this.seqMap.get(sessionId) ?? { seq: 0, lastAccess: Date.now() };
      entry.seq += 1;
      entry.lastAccess = Date.now();
      const seq = entry.seq;

      const event: TimelineEvent = {
        id: raw.id,
        sessionId,
        pageId: raw.pageId,
        ts: raw.ts ?? raw.timestamp ?? Date.now(),
        tsPage: raw.tsPage,
        type: TYPE_MAP[topic] ?? topic.replace('obs.', '').split('.')[0],
        subtype: raw.type ?? topic.split('.').pop(),
        payload: raw.payload ?? raw,
        seq,
      };

      await this.producer.send({
        topic: 'obs.timeline.merged',
        messages: [{ key: sessionId, value: JSON.stringify(event) }],
      });
    } catch (err) {
      console.error('[TimelineMerger] Error:', err);
    }
  }

  private evictStale(): void {
    const now = Date.now();
    for (const [key, entry] of this.seqMap) {
      if (now - entry.lastAccess > 30 * 60 * 1000) {
        this.seqMap.delete(key);
      }
    }
  }

  async disconnect(): Promise<void> {
    clearInterval(this.cleanupTimer);
    await this.consumer.disconnect();
    await this.producer.disconnect();
  }
}

File: api-rust/src/timeline/merger.rs
use chrono::{DateTime, Utc};
use rdkafka::message::{Message, OwnedMessage};
use rdkafka::producer::{FutureProducer, FutureRecord};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use uuid::Uuid;

use crate::metrics::KAFKA_MESSAGES_TOTAL;

#[derive(Clone)]
pub struct TimelineMerger {
    producer: FutureProducer,
    seq_counters: Arc<Mutex<HashMap<String, u64>>>,
}

impl TimelineMerger {
    pub fn new(producer: FutureProducer) -> Self {
        Self {
            producer,
            seq_counters: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn handle_message(&self, msg: &OwnedMessage) -> anyhow::Result<()> {
        let topic = msg.topic();
        let payload = msg.payload().unwrap_or_default();

        KAFKA_MESSAGES_TOTAL
            .with_label_values(&[topic, "timeline"])
            .inc();

        let mut raw: serde_json::Value = serde_json::from_slice(payload)?;

        let session_id = raw
            .get("sessionId")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| {
                msg.key()
                    .map(|k| String::from_utf8_lossy(k).to_string())
                    .unwrap_or_default()
            });

        if session_id.is_empty() {
            return Ok(());
        }

        let type_map = serde_json::json!({
            "obs.network.requests": "network",
            "obs.network.responses": "network",
            "obs.dom.mutations": "dom",
            "obs.js.events": "js",
            "obs.storage.events": "storage",
            "obs.screenshots": "screenshot",
            "obs.performance": "performance"
        });

        let ev_type = type_map
            .get(topic)
            .and_then(|v| v.as_str())
            .unwrap_or("unknown");

        let mut seq = self.seq_counters.lock().await;
        let counter = seq.entry(session_id.clone()).or_insert(0);
        *counter += 1;

        let timeline_event = serde_json::json!({
            "id": raw.get("id"),
            "sessionId": session_id,
            "pageId": raw.get("pageId"),
            "ts": raw.get("ts"),
            "tsPage": raw.get("tsPage"),
            "type": ev_type,
            "subtype": raw.get("type"),
            "payload": raw.get("payload"),
            "seq": *counter,
        });

        let payload_bytes = serde_json::to_vec(&timeline_event)?;
        let record = FutureRecord::to("obs.timeline.merged")
            .key(&session_id)
            .payload(&payload_bytes);

        self.producer.send(record, std::time::Duration::from_secs(5))
            .await
            .map_err(|(e, _)| anyhow::anyhow!("Kafka send error: {}", e))?;

        Ok(())
    }

    pub async fn cleanup_stale_sessions(&self) {
        // Sessions expire after 30 min of inactivity
        // For simplicity, clear the map periodically
        let mut seq = self.seq_counters.lock().await;
        seq.clear();
    }
}
